src/xmlparser.rs

```diff
@@ -0,0 +1,423 @@
+        self.end_element()?;
+        Ok(result)
+    }
+
+    fn start_element(&mut self) -> Result<Element> {
+        match self.next_event() {
+            Ok(XmlEvent::StartElement {name, attributes, .. }) => {
+                return Ok(Element {
+                    name,
+                    attributes,
+                    position: self.position(),
+                    error_emitter: self.error_emitter.clone(),
+                });
+            }
+            Ok(e) => Err(self.unexpected_event(&e)),
+            Err(e) => Err(e),
+        }
+    }
+
+    fn end_element(&mut self) -> Result<()> {
+        match self.next_event() {
+            Ok(XmlEvent::EndElement { .. }) => Ok(()),
+            Ok(e) => Err(self.unexpected_event(&e)),
+            Err(e) => Err(e),
+        }
+    }
+
+    pub fn text(&mut self) -> Result<String> {
+        let mut result = String::new();
+        loop {
+            match *self.peek_event() {
+                Ok(XmlEvent::Characters(..)) => {
+                    if let Ok(XmlEvent::Characters(s)) = self.next_event() {
+                        result.push_str(&s);
+                    }
+                }
+                Err(_) => {
+                    if let Err(e) = self.next_event() {
+                        return Err(e);
+                    }
+                    unreachable!();
+                }
+                _ if result.is_empty() => {
+                    return Err(self.fail("Expected text content"));
+                }
+                _ => break,
+            }
+        }
+        Ok(result)
+    }
+
+    /// Ignore everything within current element.
+    pub fn ignore_element(&mut self) -> Result<()> {
+        let mut depth = 1;
+        loop {
+            match *self.peek_event() {
+                Ok(XmlEvent::StartElement { .. }) => {
+                    // Ignore warning about unused result, we know event is OK.
+                    drop(self.next_event());
+                    depth += 1;
+                }
+                Ok(XmlEvent::EndElement { .. }) => {
+                    depth -= 1;
+                    if depth > 0 {
+                        drop(self.next_event());
+                    } else {
+                        return Ok(());
+                    }
+                }
+                Ok(_) => drop(self.next_event()),
+                Err(_) => return self.next_event().map(|_| ()),
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+
+    use super::*;
+
+    fn with_parser<F, R>(xml: &[u8], f: F) -> Result<R>
+    where
+        F: FnOnce(XmlParser) -> Result<R>,
+    {
+        f(XmlParser::new(xml)?)
+    }
+
+    #[test]
+    fn test_element_with_name() {
+        let xml = br#"<?xml version="1.0"?>
+            <!-- a comment -->
+            <secret>
+            </secret>"#;
+
+        let result = with_parser(xml, |mut p| {
+            p.document(|p, _| {
+                p.element_with_name("secret", |_, elem| {
```

Comment:
Maybe need failing test for `element_with_name`

Reply:
Done.